sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,437 @@
+        calcColOffset(i);
+      }
+      // set the validity size
+      setValidityBits(attributes.length);
+      if (hasVarSizeData()) {
+        for (int ind = varSizeColIndex; ind < attributes.length; ind++) {
+          DataType dataType = attributes[ind].dataType();
+          if (isFixedLength(dataType)) {
+            continue;
+          }
+          DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+          addVarSizeData(rapidsType);
+        }
+      }
+      byteCursor = alignOffset(byteCursor, JCUDF_ROW_ALIGNMENT);
+      return byteCursor;
+    }
+    private int addVarSizeData(DType rapidsType) {
+      int length = getEstimateSizeForVarLengthTypes(rapidsType);
+      // For variable-length, alignment and length are not equivalent.
+      int alignment = getDataAlignmentForDataType(rapidsType);
+      int res = alignOffset(byteCursor, alignment);
+      byteCursor = res + length;
+      return res;
+    }
+
+    public boolean hasVarSizeData() {
+      return attributes.length > 0 && varSizeColIndex != attributes.length;
+    }
+
+    protected int getCurrentByteOffset() {
+      return byteCursor;
+    }
+  }
+
+  /**
+   * A helper class to calculate the columns, validity, and the variable width offsets.
+   * The variable size data offset will fast-forward to calculate all offsets and cache the results
+   * for optimization purpose (trade-off is more memory usage).
+   * The size of the type is cached in {@link #typesLength}. The array will have negative values for
+   * variable-size data types.
+   * The implementation does not make any assumptions whether the schema is packed or not.
+   */
+  public static class JCudfRowVisitor {
+    Attribute[] attributes;
+    // caches the offsets of the columns.
+    int[] offsets;
+    // caches the size of the column
+    int[] typesLength;
+    int validitySizeInBytes;
+    // number of bytes used by the fixed size section (fixed fields + validity bytes)
+    int fixedWidthSizeInBytes;
+    // counter used to advance the offset as fields get added
+    int byteCursor = 0;
+    // index used to iterate through the columns.
+    int currColInd = -1;
+    // where the validity starts
+    int validityBytesOffset = 0;
+    private JCudfRowVisitor(Attribute[] attrArr) {
+      this.attributes = attrArr;
+    }
+
+    private void setValidityBits(int fieldsCount){
+      validityBytesOffset = byteCursor;
+      validitySizeInBytes = calculateBitSetWidthInBytes(fieldsCount);
+      byteCursor += validitySizeInBytes;
+    }
+
+    /**
+     * Calculate the offset of the given ind.
+     * The #typesLength caches the size of the column.
+     * If the data type is variable-size data, #typesLength will have a negative value that can be
+     * used to look-up the type. String has a length of -15.
+     *
+     * @param ind index of the column
+     * @return the offset of the colum.
+     */
+    private int calcColOffset(int ind) {
```

Comment:
This seems very familiar. Would it make sense to wrap the above class in a caching class or is it different enough?

Reply:
I was avoiding inheritance to reduce the overhead. That's why I opted to repeat the code rather than using overrides/interfaces.
In the most recent code commit, the two inner classes have different functionality.

- `JCudfUtil.RowOffsetsCalculator`: is used to calculate the offsets and estimate size for the buffer allocation.
- `JCudfUtil.RowBuilder`: it iterates on the columns to generate the code used to copy the column field.